craft_application/services/lifecycle.py

```diff
@@ -110,6 +111,7 @@ def __init__(
         cache_dir: Path | str,
```

Comment:
unrelated to this PR but this alternative to Optional needs Python >=3.10 right? Is that an issue?

Reply:
`from __future__ import annotations` brings it in. I'm less thrilled personally about changing `Optional[type]` to `type | None` than about pipes for union types and things like `list[str]`, but once you enable that mypy and complain about using `Optional`. Worth bringing up to the team what we'd like our standard to be going forward (assuming we can tell mypy/pyright to allow `Optional` if we choose).